aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,368 @@
+        return self._multiplexer.out_queue.qsize()
+
+    def get_message_from_outbox(self) -> Optional[Message]:
+        """Get message from outbox."""
+        if self._outbox.empty():
+            return None
+        envelope = self._multiplexer.out_queue.get_nowait()
+        return envelope.message
+
+    @staticmethod
+    def message_has_attributes(
+        actual_message: Message, message_type: Type[Message], **kwargs,
+    ) -> Tuple[bool, str]:
+        """
+        Evaluates whether a message's attributes match the expected attributes provided.
+
+        :param actual_message: the actual message
+        :param message_type: the expected message type
+        :param kwargs: other expected message attributes
+
+        :return: boolean result of the evaluation and accompanied message
+        """
+        if (
+            type(actual_message)  # pylint: disable=unidiomatic-typecheck
+            != message_type
+        ):
+            return (
+                False,
+                "The message types do not match. Actual type: {}. Expected type: {}".format(
+                    type(actual_message), message_type
+                ),
+            )
+
+        for attribute_name, expected_value in kwargs.items():
+            attribute = getattr(actual_message, attribute_name)
+            if callable(attribute):
+                if attribute != expected_value:
+                    return (
+                        False,
+                        f"The '{attribute_name}' fields do not match. Actual '{attribute_name}': {attribute}. Expected '{attribute_name}': {expected_value}",
+                    )
+
+        return True, "The message has the provided expected attributes."
+
+    def build_incoming_message(
+        self,
+        message_type: Type[Message],
+        performative: Message.Performative,
+        dialogue_reference: Optional[Tuple[str, str]] = None,
+        message_id: Optional[int] = None,
+        target: Optional[int] = None,
+        to: Optional[Address] = None,
+        sender: Address = "counterparty",
+        **kwargs,
+    ) -> Message:
+        """
+        Quickly create an incoming message with the provided attributes.
+
+        For any attribute not provided, the corresponding default value in message is used.
+
+        :param message_type: the type of the message
+        :param dialogue_reference: the dialogue_reference
+        :param message_id: the message_id
+        :param target: the target
+        :param performative: the performative
+        :param to: the 'to' address
+        :param sender: the 'sender' address
+        :param kwargs: other attributes
+
+        :return: the created incoming message
+        """
+        message_attributes = dict()  # type: Dict[str, Any]
+
+        default_dialogue_reference = (
+            Dialogues._generate_dialogue_nonce(),  # pylint: disable=protected-access
+            "",
+        )
```

Comment:
Would be nice if we could turn the `new_self_initiated_dialogue_reference` into a class method in the base `Dialogues` class and then call here `Dialogues.new_self_initiated_dialogue_reference()`. That should be possible without causing anything to break.

Reply:
Agree

Reply:
I don't think anything needs to be updated as a result of this change. The old usage `self.new_self_initiated_dialogue_reference()` is automatically still supported...

Reply:
Done